text/0000-feature-metadata.md

```diff
@@ -0,0 +1,275 @@
+be thought of as the equivalent to a `#[doc(...)]` attribute. Like doc comments,
+the first line should be usable as a summary, though most use cases will only
+have one line.
+
+There is nothing in this RFC that cargo `must` do with this action, since it is
+mainly intended for the consumption of `rustdoc` or `docs.rs`. However, it can
+be used for general diagnostic information.
+
+## `deprecated`
+
+`deprecated` should be thought of as the equivalent of the [`deprecated`]
+attribute in Rust source. The value can be a boolean, string, or an object with
+`since` or `note` keys. Schema rules are as follows:
+
+-   If a boolean value, `false` indicates not deprecated and `true` indicates
+    deprecated
+-   If an object, the keys `since` and/or `note` can be specified
+-   An empty object is not allowed to avoid the ambiguous
+    `foo = { deprecated = {} }`
+-   A string `foo = { deprecated = "my msg" }` will be equivalent to if that string
+    was specified in the `note` field:
+    `foo = { deprecated = { note = "my msg" } }`
+-   If not specified, the default is `false`
+
+If a downstream crate attempts to use a feature marked `deprecated`, Cargo
+should produce a warning.
+
+## `public`
+
+`public` is a boolean value that defaults to `true`. It can be thought of as
+`pub` in Rust source files, with the exception of being true by default. If set
+to `false`, Cargo should forbid its use with an error message on any downstream
+crates.
+
+There needs to be an escape hatch for this for things like benchmarks - RFC TBD
+on how this works.
+
+## `unstable`
+
+`unstable` is a boolean value that defaults to `false`. It should indicate that
+a feature gates something that is not API-stable, or can only be built using
+`nightly` Rust.
+
+---
+
+Use cases for this information will likely develop with time, but one of the
+simplest applications is for information output with `cargo add`:
+
+```text
+crab@rust foobar % cargo add regex
+    Updating crates.io index
+      Adding regex v1.7.3 to dependencies.
+             Features:
+             + perf             Enables all performance related features
+             + perf-dfa         Enables the use of a lazy DFA for matching
+             + perf-inline      Enables the use of aggressive inlining inside
+                                match routines
+             + perf-literal     Enables the use of literal optimizations for
+                                speeding up matches
+             + std              When enabled, this will cause regex to use the
+                                standard library
+             + unicode          Enables all Unicode features
+             - unstable (U)     Some unstable features
+             - deprecated (D)   Not a real feature, but it could be
+
+    Updating crates.io index
+```
+
+Features like `aho-corasick`, `memchr`, or `use_std` would likely be `public =
+false` since they aren't listed on the crate landing page.
+
+# Drawbacks
```

Comment:
There is no way to provide feature groupings with a header and body of text like with `document-features

See [clap](https://docs.rs/clap/latest/clap/_features/index.html) for a handwritten example of this

Reply:
I think the formatting outlined in that link is totally reasonable, so I added a statement about making a best effort to display features in this way (at the end of "Reference-level explanation").

Reply:
Independent of a default sort order, what I'm calling out is that developers can't customize the documentation around their features, including custom grouping

Reply:
That is a good point, I added it as a drawback https://github.com/rust-lang/rfcs/pull/3416/files#diff-b1b8fe9a7278ab777a7fa13af204b0717b1ab818716d76854ed9617569ce08d9R212-R214